File: sources/ruby/03_func_sum/main.rb
# task 03 func_sum — expected output: 100000000
# build: none (interpreted)    run: ruby main.rb (cruby) | ruby --yjit main.rb (cruby+yjit) | jruby main.rb (jruby, needs Java 25)
# The stock Windows CRuby build has no YJIT: `ruby --yjit` warns "Ruby was built without YJIT support".
# CRuby's interpreter never inlines add_one; the YJIT/JRuby JITs may inline this call site once it is hot.

def add_one(n)
  n + 1
end

value = 0
i = 0
while i < 100_000_000
  value = add_one(value)
  i += 1
end

puts value

File: sources/c/03_func_sum/main.c
// task 03 func_sum — expected output: 100000000
// build: gcc -O2 -pthread -o prog main.c    run: ./prog
// alternates: clang -O2 -pthread -o prog main.c | cl /O2 /Fe:prog main.c | tcc -o prog main.c

#include <stdio.h>
#include <stdint.h>

/* The call has to survive -O2, so the function is marked no-inline. */
#if defined(_MSC_VER)
__declspec(noinline)
#else
__attribute__((noinline))
#endif
static int64_t add_one(int64_t n) {
    return n + 1;
}

int main(void) {
    int64_t value = 0;

    for (int64_t i = 0; i < 100000000; i++) {
        value = add_one(value);
    }

    printf("%lld\n", (long long)value);
    return 0;
}
